crates/state/src/write_log.rs

```diff
@@ -1153,49 +1151,134 @@ mod tests {
-                        // Test for 2.
-                        assert!(verifiers.contains(addr_from_key));
-                    }
+        println!("verifiers {:#?}", verifiers);
+        println!("changed_keys {:#?}", changed_keys);
+        println!("initialized_accounts {:#?}", initialized_accounts);
+        for initialized_account in initialized_accounts {
+            // Test for 3.
+            assert!(!verifiers.contains(initialized_account));
+            // Test for 4.
+            let vp_key = storage::Key::validity_predicate(initialized_account);
+            assert!(changed_keys.contains(&vp_key));
+        }
+    }
+
+    /// Test that for a batched tx the result of reading modified keys from
+    /// write log matches the result of prefix iterators.
+    fn test_batched_txs_aux(
+        txs: Vec<HashMap<storage::Key, StorageModification>>,
+    ) {
+        let mut write_log = WriteLog::default();
+
+        for tx in txs {
+            // Write tx modifications
+            write_log.tx_write_log.write_log = tx;
+
+            // Collect all the keys in batch from previous txs
+            let keys: HashSet<storage::Key> = write_log
+                .batch_write_log
+                .iter()
+                .flat_map(|batch| batch.write_log.keys())
+                .cloned()
+                .collect();
+
+            // Iterate through all the modified keys to check prior state
+            for key in keys.clone() {
+                // Read the modification associated with this key from prior
+                // state
+                let (modification, _gas) = write_log.read_pre(&key).unwrap();
+                let modification = modification.unwrap();
+
+                // Prefix iter prior state for this key and assert that the
+                // values match
+                for (key_str, modification_from_iter) in
+                    write_log.iter_prefix_pre(&key)
+                {
+                    assert_eq!(key.to_string(), key_str);
+                    assert_eq!(modification, &modification_from_iter);
                 }
             }
 
-            println!("verifiers {:#?}", verifiers);
-            println!("changed_keys {:#?}", changed_keys);
-            println!("initialized_accounts {:#?}", initialized_accounts);
-            for initialized_account in initialized_accounts {
-                // Test for 3.
-                assert!(!verifiers.contains(initialized_account));
-                // Test for 4.
-                let vp_key = storage::Key::validity_predicate(initialized_account);
-                assert!(changed_keys.contains(&vp_key));
+            // And then commit them to batch
+            write_log.commit_tx_to_batch();
+
+            // Collect all the keys in batch and tx logs
+            let keys: HashSet<storage::Key> = keys
+                .into_iter()
+                .chain(write_log.tx_write_log.write_log.keys().cloned())
```

Comment:
I'm not sure if we are actually extending the keys here because `write_log.tx_write_log` is taken in `commit_tx_to_batch` to extend `write_log.batch_write_log`. Should we instead read again from `batch_write_log` like we do at line 1233?

Reply:
Or maybe we can just clone the content of `tx_write_log` before calling `commit_tx_to_batch`. We could also check that the content of `batch_write_log` after this call is the result of merging the previous value of `keys` with the content of `tx_write_log`

Reply:
ah good point, I had this originally before last force push. I'll add it back in